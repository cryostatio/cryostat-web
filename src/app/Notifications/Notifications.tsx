/*
 * Copyright The Cryostat Authors
 *
 * The Universal Permissive License (UPL), Version 1.0
 *
 * Subject to the condition set forth below, permission is hereby granted to any
 * person obtaining a copy of this software, associated documentation and/or data
 * (collectively the "Software"), free of charge and under any and all copyright
 * rights in the Software, and any and all patent rights owned or freely
 * licensable by each licensor hereunder covering either (i) the unmodified
 * Software as contributed to or provided by such licensor, or (ii) the Larger
 * Works (as defined below), to deal in both
 *
 * (a) the Software, and
 * (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
 * one is included with the Software (each a "Larger Work" to which the Software
 * is contributed by such licensors),
 *
 * without restriction, including without limitation the rights to copy, create
 * derivative works of, display, perform, and distribute the Software and make,
 * use, sell, offer for sale, import, export, have made, and have sold the
 * Software and the Larger Work(s), and to sublicense the foregoing rights on
 * either these or other terms.
 *
 * This license is subject to the following condition:
 * The above copyright notice and either this complete permission notice or at
 * a minimum a reference to the UPL must be included in all copies or
 * substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import * as React from 'react';
import { BehaviorSubject, Observable } from 'rxjs';
import { concatMap, filter, first, map } from 'rxjs/operators';
import { AlertVariant } from '@patternfly/react-core';
import { nanoid } from 'nanoid';
import { NotificationCategory } from '@app/Shared/Services/NotificationChannel.service';

export interface Notification {
  hidden?: boolean;
  read?: boolean;
  key?: string;
  title: string;
  message?: string | Error;
  category?: string;
  variant: AlertVariant;
  timestamp?: number;
}

export class Notifications {
  private readonly _notifications$: BehaviorSubject<Notification[]> = new BehaviorSubject<Notification[]>([]);
  private readonly _drawerState$: BehaviorSubject<boolean> = new BehaviorSubject<boolean>(false);

  constructor() {
    this._drawerState$
      .pipe(
        filter((v) => v),
        concatMap(() => this._notifications$.pipe(first()))
      )
      .subscribe((prev) =>
        this._notifications$.next(
          prev.map((n) => ({
            ...n,
            hidden: true,
          }))
        )
      );
  }

  drawerState(): Observable<boolean> {
    return this._drawerState$.asObservable();
  }

  setDrawerState(state: boolean): void {
    this._drawerState$.next(state);
  }

  notify(notification: Notification): void {
    if (!notification.key) {
      notification.key = nanoid();
    }
    notification.read = false;
    if (notification.hidden === undefined) {
      notification.hidden = this._drawerState$.getValue();
    }
    notification.timestamp = +Date.now();
    if (notification.message instanceof Error) {
      notification.message = JSON.stringify(notification.message, Object.getOwnPropertyNames(notification.message));
    } else if (typeof notification.message !== 'string') {
      notification.message = JSON.stringify(notification.message);
    }
    this._notifications$.pipe(first()).subscribe((prev) => {
      prev.unshift(notification);
      this._notifications$.next(prev);
    });
  }

  success(title: string, message?: string | Error, category?: string, hidden?: boolean): void {
    this.notify({ title, message, category, variant: AlertVariant.success, hidden });
  }

  info(title: string, message?: string | Error, category?: string, hidden?: boolean): void {
    this.notify({ title, message, category, variant: AlertVariant.info, hidden });
  }

  warning(title: string, message?: string | Error, category?: string): void {
    this.notify({ title, message, category, variant: AlertVariant.warning });
  }

  danger(title: string, message?: string | Error, category?: string): void {
    this.notify({ title, message, category, variant: AlertVariant.danger });
  }

  notifications(): Observable<Notification[]> {
    return this._notifications$.asObservable();
  }

  unreadNotifications(): Observable<Notification[]> {
    return this.notifications().pipe(map((a) => a.filter((n) => !n.read)));
  }

  actionsNotifications(): Observable<Notification[]> {
    return this.notifications().pipe(map((a) => a.filter((n) => this.isActionNotification(n))));
  }

  cryostatStatusNotifications(): Observable<Notification[]> {
    return this.notifications().pipe(
      map((a) =>
        a.filter(
          (n) => (this.isWsClientActivity(n) || this.isJvmDiscovery(n)) && !Notifications.isProblemNotification(n)
        )
      )
    );
  }

  problemsNotifications(): Observable<Notification[]> {
    return this.notifications().pipe(map((a) => a.filter(Notifications.isProblemNotification)));
  }

  setHidden(key?: string, hidden: boolean = true): void {
    if (!key) {
      return;
    }
    this._notifications$.pipe(first()).subscribe((prev) => {
      for (let n of prev) {
        if (n.key === key) {
          n.hidden = hidden;
        }
      }
      this._notifications$.next(prev);
    });
  }

  setRead(key?: string, read: boolean = true): void {
    if (!key) {
      return;
    }
    this._notifications$.pipe(first()).subscribe((prev) => {
      for (let n of prev) {
        if (n.key === key) {
          n.read = read;
        }
      }
      this._notifications$.next(prev);
    });
  }

  markAllRead(): void {
    this._notifications$.pipe(first()).subscribe((prev) => {
      for (let n of prev) {
        n.read = true;
      }
      this._notifications$.next(prev);
    });
  }

  clearAll(): void {
    this._notifications$.next([]);
  }

  private isActionNotification(n: Notification): boolean {
    return !this.isWsClientActivity(n) && !this.isJvmDiscovery(n) && !Notifications.isProblemNotification(n);
  }

  private isWsClientActivity(n: Notification): boolean {
    return n.category === NotificationCategory.WsClientActivity;
  }

  private isJvmDiscovery(n: Notification): boolean {
    return n.category === NotificationCategory.TargetJvmDiscovery;
  }

  static isProblemNotification(n: Notification): boolean {
    return n.variant === AlertVariant.warning || n.variant === AlertVariant.danger;
  }
}

const NotificationsInstance = new Notifications();

const NotificationsContext = React.createContext(NotificationsInstance);

export { NotificationsContext, NotificationsInstance };
