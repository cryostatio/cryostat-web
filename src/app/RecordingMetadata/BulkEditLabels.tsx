/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { uploadAsTarget } from '@app/Archives/Archives';
import { LoadingProps } from '@app/Shared/Components/types';
import {
  RecordingDirectory,
  ArchivedRecording,
  Recording,
  ActiveRecording,
  UPLOADS_SUBDIRECTORY,
  NotificationCategory,
  Target,
  KeyValue,
} from '@app/Shared/Services/api.types';
import { ServiceContext } from '@app/Shared/Services/Services';
import { useSubscriptions } from '@app/utils/hooks/useSubscriptions';
import { hashCode } from '@app/utils/utils';
import {
  ActionList,
  ActionListItem,
  Button,
  HelperText,
  HelperTextItem,
  Label,
  Stack,
  StackItem,
  Title,
  ValidatedOptions,
} from '@patternfly/react-core';
import * as React from 'react';
import { combineLatest, concatMap, filter, first, forkJoin, map, Observable, of } from 'rxjs';
import { RecordingLabelFields } from './RecordingLabelFields';
import { includesLabel } from './utils';

export interface BulkEditLabelsProps {
  isTargetRecording: boolean;
  isUploadsTable?: boolean;
  checkedIndices: number[];
  directory?: RecordingDirectory;
  directoryRecordings?: ArchivedRecording[];
  closePanelFn?: () => void;
}

export const BulkEditLabels: React.FC<BulkEditLabelsProps> = ({
  isTargetRecording,
  isUploadsTable,
  checkedIndices,
  directory,
  directoryRecordings,
  closePanelFn,
}) => {
  const context = React.useContext(ServiceContext);
  const [recordings, setRecordings] = React.useState<Recording[]>([]);
  const [commonLabels, setCommonLabels] = React.useState<KeyValue[]>([]);
  const [savedCommonLabels, setSavedCommonLabels] = React.useState<KeyValue[]>([]);
  const [valid, setValid] = React.useState(ValidatedOptions.default);
  const [loading, setLoading] = React.useState(false);
  const addSubscription = useSubscriptions();

  const getIdxFromRecording = React.useCallback(
    (r: Recording): number => (isTargetRecording ? (r as ActiveRecording).id : hashCode(r.name)),
    [isTargetRecording],
  );

  const handlePostUpdate = React.useCallback(() => {
    setLoading(false);
  }, [setLoading]);

  const handleUpdateLabels = React.useCallback(() => {
    setLoading(true);
    const tasks: Observable<unknown>[] = [];
    const toDelete = savedCommonLabels.filter((label) => !includesLabel(commonLabels, label));

    recordings.forEach((r: Recording) => {
      const idx = getIdxFromRecording(r);
      if (checkedIndices.includes(idx)) {
        const updatedLabels = [...r.metadata.labels, ...commonLabels].filter(
          (label) => !includesLabel(toDelete, label),
        );

        if (directory) {
          tasks.push(context.api.postRecordingMetadataForJvmId(directory.jvmId, r.name, updatedLabels).pipe(first()));
        }
        if (isTargetRecording) {
          tasks.push(context.api.postTargetRecordingMetadata(r.name, updatedLabels).pipe(first()));
        } else if (isUploadsTable) {
          tasks.push(context.api.postUploadedRecordingMetadata(r.name, updatedLabels).pipe(first()));
        } else {
          tasks.push(context.api.postRecordingMetadata(r.name, updatedLabels).pipe(first()));
        }
      }
    });
    addSubscription(
      forkJoin(tasks).subscribe({
        next: handlePostUpdate,
        error: handlePostUpdate,
      }),
    );
  }, [
    addSubscription,
    context.api,
    getIdxFromRecording,
    handlePostUpdate,
    commonLabels,
    savedCommonLabels,
    recordings,
    checkedIndices,
    isTargetRecording,
    isUploadsTable,
    directory,
  ]);

  const handleCancel = React.useCallback(() => {
    setCommonLabels(savedCommonLabels);
    closePanelFn && closePanelFn();
  }, [setCommonLabels, savedCommonLabels, closePanelFn]);

  const updateCommonLabels = React.useCallback(
    (setLabels: (l: KeyValue[]) => void) => {
      const allRecordingLabels: KeyValue[][] = [];

      recordings.forEach((r: Recording) => {
        const idx = getIdxFromRecording(r);
        if (checkedIndices.includes(idx)) {
          allRecordingLabels.push(r.metadata.labels);
        }
      });

      const updatedCommonLabels =
        allRecordingLabels.length > 0
          ? allRecordingLabels.reduce(
              (prev, curr) => prev.filter((label) => includesLabel(curr, label)),
              allRecordingLabels[0],
            )
          : [];

      setLabels(updatedCommonLabels);
    },
    [recordings, getIdxFromRecording, checkedIndices],
  );

  /* eslint-disable @typescript-eslint/no-explicit-any */
  const refreshRecordingList = React.useCallback(() => {
    let observable: Observable<Recording[]>;
    if (directoryRecordings) {
      observable = of(directoryRecordings);
    } else if (isTargetRecording) {
      observable = context.target.target().pipe(
        filter((target) => !!target),
        concatMap((target: Target) => context.api.getTargetActiveRecordings(target)),
        first(),
      );
    } else {
      observable = isUploadsTable
        ? context.api
            .graphql<any>(
              `query GetUploadedRecordings($filter: ArchivedRecordingsFilterInput) {
                archivedRecordings(filter: $filter) {
                  data {
                    name
                    downloadUrl
                    reportUrl
                    metadata {
                      labels {
                        key
                        value
                      }
                    }
                  }
                }
              }`,
              { filter: { sourceTarget: UPLOADS_SUBDIRECTORY } },
            )
            .pipe(
              map((v) => (v.data?.archivedRecordings?.data as ArchivedRecording[]) ?? []),
              first(),
            )
        : context.target.target().pipe(
            filter((target) => !!target),
            concatMap((target: Target) =>
              context.api.graphql<any>(
                `query ArchivedRecordingsForTarget($id: BigInteger!) {
                targetNodes(filter: { targetIds: [$id] }) {
                  target {
                    archivedRecordings {
                      data {
                        name
                        downloadUrl
                        reportUrl
                        metadata {
                          labels {
                            key
                            value
                          }
                        }
                        size
                        archivedTime
                      }
                    }
                  }
                }
              }`,
                { id: target.id! },
              ),
            ),
            map((v) => (v.data?.targetNodes[0]?.target?.archivedRecordings?.data as ArchivedRecording[]) ?? []),
            first(),
          );
    }

    addSubscription(observable.subscribe((value) => setRecordings(value)));
  }, [
    addSubscription,
    isTargetRecording,
    isUploadsTable,
    directoryRecordings,
    context.target,
    context.api,
    setRecordings,
  ]);
  /* eslint-enable @typescript-eslint/no-explicit-any */

  const saveButtonLoadingProps = React.useMemo(
    () =>
      ({
        spinnerAriaValueText: 'Saving',
        spinnerAriaLabel: 'saving-recording-labels',
        isLoading: loading,
      }) as LoadingProps,
    [loading],
  );

  React.useEffect(() => {
    addSubscription(context.target.target().subscribe(refreshRecordingList));
  }, [addSubscription, context, context.target, refreshRecordingList]);

  // Depends only on RecordingMetadataUpdated notifications
  // since updates on list of recordings will mount a completely new BulkEditLabels.
  React.useEffect(() => {
    addSubscription(
      combineLatest([
        isUploadsTable ? of(uploadAsTarget) : context.target.target(),
        context.notificationChannel.messages(NotificationCategory.RecordingMetadataUpdated),
      ]).subscribe((parts) => {
        const currentTarget = parts[0];
        const event = parts[1];

        const isMatch =
          currentTarget?.jvmId === event.message.jvmId ||
          currentTarget?.jvmId === event.message.recording.jvmId ||
          currentTarget?.jvmId === 'uploads';

        setRecordings((oldRecordings) => {
          return oldRecordings.map((recording) => {
            if (isMatch && recording.name === event.message.recording.name) {
              const updatedRecording = {
                ...recording,
                metadata: {
                  labels: event.message.recording.metadata.labels,
                },
              };
              return updatedRecording;
            }
            return recording;
          });
        });
      }),
    );
  }, [addSubscription, context.target, context.notificationChannel, setRecordings, isUploadsTable]);

  React.useEffect(() => {
    updateCommonLabels(setCommonLabels);
    updateCommonLabels(setSavedCommonLabels);
  }, [recordings, setCommonLabels, setSavedCommonLabels, updateCommonLabels]);

  return (
    <>
      <Stack hasGutter>
        <StackItem>
          <Title headingLevel="h2">Edit labels</Title>
        </StackItem>
        <StackItem>
          <HelperText>
            <HelperTextItem>
              Labels present on all selected Recordings will appear here. Editing the labels will affect all selected
              Recordings. Specify labels with format <Label isCompact>key=value</Label>.
            </HelperTextItem>
          </HelperText>
        </StackItem>
        <StackItem>
          <RecordingLabelFields
            labels={commonLabels}
            setLabels={setCommonLabels}
            setValid={setValid}
            isDisabled={loading}
          />
        </StackItem>
        <StackItem>
          <ActionList>
            <ActionListItem>
              <Button
                variant="primary"
                onClick={handleUpdateLabels}
                isDisabled={valid != ValidatedOptions.success || loading}
                {...saveButtonLoadingProps}
              >
                {loading ? 'Saving' : 'Save'}
              </Button>
            </ActionListItem>
            <ActionListItem>
              <Button variant="secondary" onClick={handleCancel} isDisabled={loading}>
                Cancel
              </Button>
            </ActionListItem>
          </ActionList>
        </StackItem>
      </Stack>
    </>
  );
};
