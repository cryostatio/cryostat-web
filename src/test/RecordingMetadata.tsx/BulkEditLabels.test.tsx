/*
 * Copyright The Cryostat Authors
 *
 * The Universal Permissive License (UPL), Version 1.0
 *
 * Subject to the condition set forth below, permission is hereby granted to any
 * person obtaining a copy of this software, associated documentation and/or data
 * (collectively the "Software"), free of charge and under any and all copyright
 * rights in the Software, and any and all patent rights owned or freely
 * licensable by each licensor hereunder covering either (i) the unmodified
 * Software as contributed to or provided by such licensor, or (ii) the Larger
 * Works (as defined below), to deal in both
 *
 * (a) the Software, and
 * (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
 * one is included with the Software (each a "Larger Work" to which the Software
 * is contributed by such licensors),
 *
 * without restriction, including without limitation the rights to copy, create
 * derivative works of, display, perform, and distribute the Software and make,
 * use, sell, offer for sale, import, export, have made, and have sold the
 * Software and the Larger Work(s), and to sublicense the foregoing rights on
 * either these or other terms.
 *
 * This license is subject to the following condition:
 * The above copyright notice and either this complete permission notice or at
 * a minimum a reference to the UPL must be included in all copies or
 * substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import * as React from 'react';
import renderer, { act } from 'react-test-renderer';
import { cleanup, screen } from '@testing-library/react';
import { of } from 'rxjs';
import '@testing-library/jest-dom';
import { BulkEditLabels } from '@app/RecordingMetadata/BulkEditLabels';
import { ServiceContext, defaultServices } from '@app/Shared/Services/Services';
import { ActiveRecording, ArchivedRecording, RecordingState } from '@app/Shared/Services/Api.service';
import { NotificationMessage } from '@app/Shared/Services/NotificationChannel.service';
import { renderWithServiceContext } from '../Common';
import { NotificationsContext, NotificationsInstance } from '@app/Notifications/Notifications';

jest.mock('@patternfly/react-core', () => ({
  ...jest.requireActual('@patternfly/react-core'),
  Tooltip: ({ t }) => <>{t}</>,
}));

const mockConnectUrl = 'service:jmx:rmi://someUrl';
const mockTarget = { connectUrl: mockConnectUrl, alias: 'fooTarget' };

const mockRecordingLabels = {
  someLabel: 'someValue',
};

const mockArchivedRecording: ArchivedRecording = {
  name: 'someArchivedRecording_some_random',
  downloadUrl: 'http://downloadUrl',
  reportUrl: 'http://reportUrl',
  metadata: { labels: mockRecordingLabels },
  size: 2048,
  archivedTime: 2048,
};

const mockActiveRecording: ActiveRecording = {
  name: 'someActiveRecording',
  downloadUrl: 'http://downloadUrl',
  reportUrl: 'http://reportUrl',
  metadata: { labels: mockRecordingLabels },
  startTime: 1234567890,
  id: 0,
  state: RecordingState.RUNNING,
  duration: 0,
  continuous: false,
  toDisk: false,
  maxSize: 0,
  maxAge: 0,
};

const mockActiveLabelsNotification = {
  message: {
    target: mockConnectUrl,
    recordingName: 'someActiveRecording',
    metadata: { labels: { someLabel: 'someValue', someNewLabel: 'someNewValue' } },
  },
} as NotificationMessage;

const mockActiveRecordingResponse = [mockActiveRecording];

const mockArchivedLabelsNotification = {
  message: {
    target: mockConnectUrl,
    recordingName: 'someArchivedRecording_some_random',
    metadata: { labels: { someLabel: 'someValue', someNewLabel: 'someNewValue' } },
  },
} as NotificationMessage;

const mockArchivedRecordingsResponse = {
  data: {
    targetNodes: [
      {
        recordings: {
          archived: {
            data: [mockArchivedRecording] as ArchivedRecording[],
          },
        },
      },
    ],
  },
};

jest.spyOn(defaultServices.target, 'target').mockReturnValue(of(mockTarget));
jest.spyOn(defaultServices.api, 'graphql').mockReturnValue(of(mockArchivedRecordingsResponse));
jest.spyOn(defaultServices.api, 'doGet').mockReturnValue(of(mockActiveRecordingResponse));
jest
  .spyOn(defaultServices.notificationChannel, 'messages')
  .mockReturnValueOnce(of()) // renders correctly
  .mockReturnValueOnce(of()) // should display read-only labels from selected recordings
  .mockReturnValueOnce(of()) // should display editable labels form when in edit mode
  .mockReturnValueOnce(of()) // should not display labels for unchecked recordings
  .mockReturnValueOnce(of(mockActiveLabelsNotification)) // should update the target recording labels after receiving a notification
  .mockReturnValueOnce(of(mockArchivedLabelsNotification)) // should update the archived recording labels after receiving a notification
  .mockReturnValue(of());

describe('<BulkEditLabels />', () => {
  let activeCheckedIndices: number[];
  let archivedCheckedIndices: number[];
  let emptycheckIndices: number[];

  beforeEach(() => {
    activeCheckedIndices = [mockActiveRecording.id];
    archivedCheckedIndices = [-553224758]; // Hash code of "someArchivedRecording_some_random"
    emptycheckIndices = [];
  });

  afterEach(cleanup);

  it('renders correctly', async () => {
    let tree;
    await act(async () => {
      tree = renderer.create(
        <ServiceContext.Provider value={defaultServices}>
          <NotificationsContext.Provider value={NotificationsInstance}>
            <BulkEditLabels checkedIndices={activeCheckedIndices} isTargetRecording={true} />
          </NotificationsContext.Provider>
        </ServiceContext.Provider>
      );
    });
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('should display read-only labels from selected recordings', async () => {
    renderWithServiceContext(<BulkEditLabels checkedIndices={activeCheckedIndices} isTargetRecording={true} />);

    const label = screen.getByLabelText('someLabel: someValue');
    expect(label).toBeInTheDocument();
    expect(label).toBeVisible();
    expect(label.onclick).toBeNull();

    const addLabelButton = screen.queryByRole('button', { name: 'Add Label' });
    expect(addLabelButton).not.toBeInTheDocument();

    const editButton = screen.getByRole('button', { name: 'Edit Labels' });
    expect(editButton).toBeInTheDocument();
    expect(editButton).toBeVisible();
    expect(editButton).not.toBeDisabled();
  });

  it('should not display labels for unchecked recordings', async () => {
    renderWithServiceContext(<BulkEditLabels checkedIndices={emptycheckIndices} isTargetRecording={true} />);

    expect(screen.queryByText('someLabel')).not.toBeInTheDocument();
    expect(screen.queryByText('someValue')).not.toBeInTheDocument();

    const placeHolder = screen.getByText('-');
    expect(placeHolder).toBeInTheDocument();
    expect(placeHolder).toBeVisible();
  });

  it('should display editable labels form when in edit mode', async () => {
    const { user } = renderWithServiceContext(
      <BulkEditLabels checkedIndices={activeCheckedIndices} isTargetRecording={true} />
    );

    const editButton = screen.getByRole('button', { name: 'Edit Labels' });
    expect(editButton).toBeInTheDocument();
    expect(editButton).toBeVisible();

    await user.click(editButton);

    const addLabelButton = screen.getByRole('button', { name: 'Add Label' });
    expect(addLabelButton).toBeInTheDocument();
    expect(addLabelButton).toBeVisible();

    const labelKeyInput = screen.getAllByDisplayValue('someLabel')[0];
    expect(labelKeyInput).toHaveClass('pf-c-form-control');
    expect(labelKeyInput).toHaveClass('pf-c-form-control');
    expect(labelKeyInput).toBeInTheDocument();
    expect(labelKeyInput).toBeVisible();

    const labelValueInput = screen.getAllByDisplayValue('someValue')[0];
    expect(labelValueInput).toHaveClass('pf-c-form-control');
    expect(labelValueInput).toHaveClass('pf-c-form-control');
    expect(labelValueInput).toBeInTheDocument();
    expect(labelValueInput).toBeVisible();

    const saveButton = screen.getByText('Save');
    expect(saveButton).toBeInTheDocument();
    expect(saveButton).toBeVisible();

    const cancelButton = screen.getByText('Cancel');
    expect(cancelButton).toBeInTheDocument();
    expect(cancelButton).toBeVisible();
  });

  it('should update the target recording labels after receiving a notification', async () => {
    renderWithServiceContext(<BulkEditLabels checkedIndices={activeCheckedIndices} isTargetRecording={true} />);

    const newLabel = screen.getByLabelText('someNewLabel: someNewValue');
    expect(newLabel).toBeInTheDocument();
    expect(newLabel).toBeVisible();

    const oldLabel = screen.getByLabelText('someLabel: someValue');
    expect(oldLabel).toBeInTheDocument();
    expect(oldLabel).toBeVisible();
  });

  it('should update the archived recording labels after receiving a notification', async () => {
    renderWithServiceContext(<BulkEditLabels checkedIndices={archivedCheckedIndices} isTargetRecording={false} />);

    const newLabel = screen.getByLabelText('someNewLabel: someNewValue');
    expect(newLabel).toBeInTheDocument();
    expect(newLabel).toBeVisible();

    const oldLabel = screen.getByLabelText('someLabel: someValue');
    expect(oldLabel).toBeInTheDocument();
    expect(oldLabel).toBeVisible();
  });

  it('should return to read-only view when edited labels are cancelled', async () => {
    const { user } = renderWithServiceContext(
      <BulkEditLabels checkedIndices={archivedCheckedIndices} isTargetRecording={false} />
    );

    let editButton = screen.getByRole('button', { name: 'Edit Labels' });
    expect(editButton).toBeInTheDocument();
    expect(editButton).toBeVisible();

    await user.click(editButton);

    let addLabelButton = screen.getByRole('button', { name: 'Add Label' });
    expect(addLabelButton).toBeInTheDocument();
    expect(addLabelButton).toBeVisible();

    const labelKeyInput = screen.getAllByDisplayValue('someLabel')[0];
    expect(labelKeyInput).toHaveClass('pf-c-form-control');
    expect(labelKeyInput).toHaveClass('pf-c-form-control');
    expect(labelKeyInput).toBeInTheDocument();
    expect(labelKeyInput).toBeVisible();

    const labelValueInput = screen.getAllByDisplayValue('someValue')[0];
    expect(labelValueInput).toHaveClass('pf-c-form-control');
    expect(labelValueInput).toHaveClass('pf-c-form-control');
    expect(labelValueInput).toBeInTheDocument();
    expect(labelValueInput).toBeVisible();

    const saveButton = screen.getByText('Save');
    expect(saveButton).toBeInTheDocument();
    expect(saveButton).toBeVisible();

    const cancelButton = screen.getByText('Cancel');
    expect(cancelButton).toBeInTheDocument();
    expect(cancelButton).toBeVisible();

    await user.click(cancelButton);

    editButton = screen.getByRole('button', { name: 'Edit Labels' });
    expect(editButton).toBeInTheDocument();
    expect(editButton).toBeVisible();
    expect(addLabelButton).not.toBeInTheDocument();
  });

  it('should save target recording labels when Save is clicked', async () => {
    const saveRequestSpy = jest
      .spyOn(defaultServices.api, 'postTargetRecordingMetadata')
      .mockReturnValue(of([mockActiveRecording]));
    const { user } = renderWithServiceContext(
      <BulkEditLabels checkedIndices={activeCheckedIndices} isTargetRecording={true} />
    );

    let editButton = screen.getByRole('button', { name: 'Edit Labels' });
    expect(editButton).toBeInTheDocument();
    expect(editButton).toBeVisible();

    await user.click(editButton);

    let addLabelButton = screen.getByRole('button', { name: 'Add Label' });
    expect(addLabelButton).toBeInTheDocument();
    expect(addLabelButton).toBeVisible();

    const labelKeyInput = screen.getAllByDisplayValue('someLabel')[0];
    expect(labelKeyInput).toHaveClass('pf-c-form-control');
    expect(labelKeyInput).toHaveClass('pf-c-form-control');
    expect(labelKeyInput).toBeInTheDocument();
    expect(labelKeyInput).toBeVisible();

    const labelValueInput = screen.getAllByDisplayValue('someValue')[0];
    expect(labelValueInput).toHaveClass('pf-c-form-control');
    expect(labelValueInput).toHaveClass('pf-c-form-control');
    expect(labelValueInput).toBeInTheDocument();
    expect(labelValueInput).toBeVisible();

    const saveButton = screen.getByText('Save');
    expect(saveButton).toBeInTheDocument();
    expect(saveButton).toBeVisible();

    const cancelButton = screen.getByText('Cancel');
    expect(cancelButton).toBeInTheDocument();
    expect(cancelButton).toBeVisible();

    await user.click(saveButton);

    expect(saveRequestSpy).toHaveBeenCalledTimes(1);
  });

  it('should save archived recording labels when Save is clicked', async () => {
    const saveRequestSpy = jest
      .spyOn(defaultServices.api, 'postRecordingMetadata')
      .mockReturnValue(of([mockArchivedRecording]));
    const { user } = renderWithServiceContext(
      <BulkEditLabels checkedIndices={archivedCheckedIndices} isTargetRecording={false} />
    );

    let editButton = screen.getByRole('button', { name: 'Edit Labels' });
    expect(editButton).toBeInTheDocument();
    expect(editButton).toBeVisible();

    await user.click(editButton);

    let addLabelButton = screen.getByRole('button', { name: 'Add Label' });
    expect(addLabelButton).toBeInTheDocument();
    expect(addLabelButton).toBeVisible();

    const labelKeyInput = screen.getAllByDisplayValue('someLabel')[0];
    expect(labelKeyInput).toHaveClass('pf-c-form-control');
    expect(labelKeyInput).toHaveClass('pf-c-form-control');
    expect(labelKeyInput).toBeInTheDocument();
    expect(labelKeyInput).toBeVisible();

    const labelValueInput = screen.getAllByDisplayValue('someValue')[0];
    expect(labelValueInput).toHaveClass('pf-c-form-control');
    expect(labelValueInput).toHaveClass('pf-c-form-control');
    expect(labelValueInput).toBeInTheDocument();
    expect(labelValueInput).toBeVisible();

    const saveButton = screen.getByText('Save');
    expect(saveButton).toBeInTheDocument();
    expect(saveButton).toBeVisible();

    const cancelButton = screen.getByText('Cancel');
    expect(cancelButton).toBeInTheDocument();
    expect(cancelButton).toBeVisible();

    await user.click(saveButton);

    expect(saveRequestSpy).toHaveBeenCalledTimes(1);
  });
});
