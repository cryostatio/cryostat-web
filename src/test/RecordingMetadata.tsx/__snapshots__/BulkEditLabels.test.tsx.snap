// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BulkEditLabels /> renders correctly 1`] = `
<div
  className="pf-l-stack pf-m-gutter"
>
  <div
    className="pf-l-stack__item"
  >
    <p
      className=""
      data-ouia-component-id="OUIA-Generated-Text-1"
      data-ouia-component-type="PF4/Text"
      data-ouia-safe={true}
      data-pf-content={true}
    >
      Edit Recording Labels
    </p>
    <small
      className=""
      data-ouia-component-id="OUIA-Generated-Text-2"
      data-ouia-component-type="PF4/Text"
      data-ouia-safe={true}
      data-pf-content={true}
    >
      Labels present on all selected recordings will appear here. Editing the labels will affect ALL selected recordings.
    </small>
  </div>
  <div
    className="pf-l-stack__item"
  >
    <span
      className="pf-c-label"
    >
      <span
        className="pf-c-label__content"
      >
        someLabel: someValue
      </span>
    </span>
  </div>
  <div
    className="pf-l-stack__item"
  >
    <button
      aria-disabled={false}
      aria-label={null}
      className="pf-c-button pf-m-secondary"
      data-ouia-component-id="OUIA-Generated-Button-secondary-1"
      data-ouia-component-type="PF4/Button"
      data-ouia-safe={true}
      disabled={false}
      onClick={[Function]}
      role={null}
      type="button"
    >
      Edit
    </button>
  </div>
</div>
`;
